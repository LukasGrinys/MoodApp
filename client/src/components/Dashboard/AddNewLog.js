import React, {Component} from 'react';
import FontAwesome from 'react-fontawesome'
import { connect } from 'react-redux';
import { canUserLog } from '../../actions';
import styles from './AddNewLog.module.scss';
import LoadingNetItem from '../Loading/loadingNetItem';
import { Link } from 'react-router-dom';
import { getDate, getDaytime } from '../../util/dateHelpers';
import { daytimes } from '../../constants/daytimes';
import { routerRoutes } from '../../constants/routerRoutes';
import classNames from 'classnames';

class AddNewLog extends Component {
    componentWillMount() {
        const date = getDate();
        const timing = getDaytime();
        this.props.dispatch(canUserLog(date, timing, this.props.userId))
    };

    returnTimeRemaining() {
        let currentTiming = getDaytime();
        let timeOfNextLog = 0;
        
        switch(currentTiming) {
            case daytimes.morning:
                timeOfNextLog = 13;
                break;
            case daytimes.afternoon :
                timeOfNextLog = 20;
                break;
            case daytimes.evening :
                timeOfNextLog = 7;
                break;
            default:
                timeOfNextLog = 7;
        };

        return `${timeOfNextLog}:00`
    }

    render(){
        const { canLog, darkTheme } = this.props;

        if (canLog === undefined) {
            return <LoadingNetItem/>
        } 

        if (canLog && getDaytime() !== daytimes.night) {
            return (
                <Link to={routerRoutes.addNewLog}>
                    <div className={classNames(
                        styles.logbutton, 
                        styles.available,
                        darkTheme && styles.dark
                        )}>
                        <div className={styles.attentionBorder}/>
                        <FontAwesome name="fas fa-plus" className={styles.floating}/> ADD LOG
                    </div>
                </Link> 
            )
        }

        return (
            <div className={classNames(styles.logbutton, styles.unavailable)}>
                Next log will be available at: {this.returnTimeRemaining()}
            </div>
        )
    }
};

function mapStateToProps({logs, user}) {
    const { id : userId } = user;
    const { canLog } = logs; 

    return { canLog, userId }
}

export default connect(mapStateToProps)(AddNewLog);