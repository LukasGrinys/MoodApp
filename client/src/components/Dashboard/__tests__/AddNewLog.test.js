import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import AddNewLog from '../AddNewLog';
import LoadingNetItem from '../../Loading/loadingNetItem';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { returnTimeRemaining } from '../AddNewLog';
const mockStore = configureStore();

describe('AddNewLog button: Loading', () => {
    const initialState = {
        user: {},
        logs: {
            canLog : null
        }
    }

    const store = mockStore(initialState);

    const wrapper = mount(
        <Provider store={store}>
            <AddNewLog/>
        </Provider>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has LoadingNetItem widget while loading', () => {
        expect(wrapper.find(LoadingNetItem).length).toBe(1);
    });
});

describe('AddNewLog button: Data loaded', () => {
    test('Renders properly: user can log', () => {
        const mockedStore = mockStore({
            user: {},
            logs: {
                canLog : true
            }
        });

        const wrapper = mount(
            <Router>
                <Provider store={mockedStore}>
                    <AddNewLog />
                </Provider>
            </Router>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('Renders properly: user cant log', () => {
        const mockedStore = mockStore({
            user: {},
            logs: {
                canLog : false
            }
        });

        const wrapper = mount(
            <Provider store={mockedStore}>
                <AddNewLog/>
            </Provider>
        );

        const userCantLogText = `Next log will be available at: ${returnTimeRemaining()}`;
        expect(wrapper.find('.logbutton').text()).toBe(userCantLogText);
    });
})