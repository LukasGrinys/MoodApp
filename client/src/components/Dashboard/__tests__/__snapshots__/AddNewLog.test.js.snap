// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewLog button: Data loaded Renders properly: user can log 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(AddNewLog)>
        <AddNewLog
          canLog={true}
          dispatch={[Function]}
          userId={null}
        >
          <Link
            to="/addlog"
          >
            <LinkAnchor
              href="/addlog"
              navigate={[Function]}
            >
              <a
                href="/addlog"
                onClick={[Function]}
              >
                <div
                  className="logbutton available"
                >
                  <div
                    className="attentionBorder"
                  />
                  <FontAwesome
                    className="floating"
                    name="fas fa-plus"
                  >
                    <span
                      aria-hidden={true}
                      className="fa fa-fas fa-plus floating"
                    />
                  </FontAwesome>
                   ADD LOG
                </div>
              </a>
            </LinkAnchor>
          </Link>
        </AddNewLog>
      </Connect(AddNewLog)>
    </Provider>
  </Router>
</BrowserRouter>
`;

exports[`AddNewLog button: Loading Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(AddNewLog)>
    <AddNewLog
      canLog={null}
      dispatch={[Function]}
      userId={null}
    >
      <LoadingNetItem>
        <div
          className="loadingScreen"
        >
          <div
            className="barContainer"
          >
            <div
              className="bar"
            />
            <div
              className="bar delay1"
            />
            <div
              className="bar delay2"
            />
          </div>
        </div>
      </LoadingNetItem>
    </AddNewLog>
  </Connect(AddNewLog)>
</Provider>
`;
